File: ebook/code/js/chapter02/chapter02.js
/*
 * By Frederick-S
 * 2014-01-23
 */

// 1. 暴力轮询

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare1('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare1(a, b) {
    var i, j, m = b.length, n = a.length;

    for (i = 0; i < m; i++) {
        for (j = 0; j < n && a[j] != b[i]; j++)
            ;

        if (j == n) {
            return false;
        }
    };

    return true;
}

// 2. 普通排序

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare2('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare2(a, b) {
    a = a.split('').sort();
    b = b.split('').sort();

    for (var i = 0, j = 0, m = b.length, n = a.length; j < m; j++) {
        for (; i < n && a[i] < b[j]; i++)
            ;

        if (i == n || a[i] > b[j]) {
            return false;
        }
    }

    return true;
}

// 3. 计数比较

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare3('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare3(a, b) {
    var occurrences = new Array(26), m = b.length, n = a.length, index;

    for (var i = 0; i < n; i++) {
        index = a[i].charCodeAt(0) - 'A'.charCodeAt(0);
        occurrences[index]++ || (occurrences[index] = 1);
    }

    for (var j = 0; j < m; j++) {
        index = b[j].charCodeAt(0) - 'A'.charCodeAt(0);
        if (!occurrences[index]) {
            return false;
        }
    }

    return true;
}

// 4. 巧用 hashtable

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare4('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare4(a, b) {
    var obj = {}, k = 0, m = b.length, n = a.length, index;

    for (var i = 0; i < m; i++) {
        if (!obj[b[i]]) {
            obj[b[i]] = 1;
            k++;
        }
    }

    for (var j = 0; j < n; j++) {
        if (obj[a[j]]) {
            k--;
            obj[a[j]] = 0;
        }
    }

    return k === 0;
}

// 5. 素数相乘
// 此方法只有理论意义, 因为整数乘积很大, 会溢出

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare5('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare5(a, b) {
    var primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101], m = b.length, n = a.length, p = 1, t, index;

    for (var i = 0; i < n; i++) {
        index = a[i].charCodeAt(0) - 'A'.charCodeAt(0);
        t = primes[index];

        if (p % t) {
            p *= t;
        }
    }

    for (var j = 0; j < m; j++) {
        index = b[j].charCodeAt(0) - 'A'.charCodeAt(0);
        t = primes[index];

        if (p % t) {
            return false;
        }
    }

    return true;
}

// 6. 位运算

/*
 * @param {String} a 较长的字符串
 * @param {Number} b 较短的字符串
 * @example
 * var result = compare6('ABCDEFGHLMNOPQRS', 'DCGSRQPO');
 */
function compare6(a, b) {
    var hash = 0, m = b.length, n = a.length, index;

    for (var i = 0; i < n; i++) {
        index = a[i].charCodeAt(0) - 'A'.charCodeAt(0);
        hash |= 1 << index;
    }

    for (var j = 0; j < m; j++) {
        index = b[j].charCodeAt(0) - 'A'.charCodeAt(0);
        if ((hash & (1 << index)) === 0) {
            return false;
        }
    }

    return true;
}

File: ebook/code/ruby/chapter02.rb
#!/usr/bin/env ruby
# -*- coding: UTF-8 -*-

def compare_1(long_string, short_string)
  short_string.each_char{|s| 
    return false unless long_string.include? s
  }
  true
end

def compare_2(long_string, short_string)
  long_string = long_string.chars.sort.join
  short_string = short_string.chars.sort.join
  i,j, n, m = 0, 0, long_string.length, short_string.length
  while i < n && j < m
    i += 1 while long_string[i] < short_string[j] && i < n - 1
    break unless long_string[i] == short_string[j]
    j += 1
  end
  j == m
end

def count_sort(src)
  dst = [""] * src.length
  tmp = [0] * 26
  src.each_char{| c|   tmp[ c.ord - "A".ord ] += 1}
  (1...(tmp.length)).each{ |i| tmp[i] += tmp[i - 1]}
  (src.length - 1).downto(0){ |i|
    c = src[i]
    index = c.ord - "A".ord
    dst[tmp[index] - 1] = c
    tmp[index] -= 1
  }
  dst.join
end


def compare_3(long_string, short_string)
  sorted_long = count_sort(long_string)
  sorted_short = count_sort(short_string)
  pos_long, pos_short = 0, 0
  while pos_short < sorted_short.length && pos_long < sorted_long.length
    while (sorted_long[pos_long] < sorted_short[pos_short] && pos_long < sorted_long.length - 1)
      pos_long += 1
    end
    break if (sorted_long[pos_long] != sorted_short[pos_short])
    pos_short += 1
    
  end

  pos_short == sorted_short.length
end

#这个方法依赖ruby 2.0以上版本, 原因在 String.ord是在ruby 2.0以上才有
#在2.0以前版本可以使用 'A'.unpack('C')[0] 来获得ASCII
def compare_4(long_string, short_string)
  hash_temp , count = [0] * 26, 0
  short_string.each_char{ |c|
    index = c.ord - "A".ord 
    if hash_temp[ index ]  == 0
      hash_temp[index] += 1
      count += 1
    end
  }
  long_string.each_char{ |c|
    index = c.ord - "A".ord 
    if(hash_temp[index] == 1)
      hash_temp[index] -= 1
      count -= 1
    end
  }
  count == 0
end

#这个方法依赖ruby 2.0以上版本, 原因在 String.ord是在ruby 2.0以上才有
#在2.0以前版本可以使用 'A'.unpack('C')[0] 来获得ASCII
def compare_5(long_string, short_string)
  hash_temp = [false] * 26
  short_string.each_char{ |c|
    hash_temp[ c.ord - "A".ord ] = true
  }
  long_string.each_char{ |c|
    hash_temp[ c.ord - "A".ord ] = false
  }
  !hash_temp.include? true
end

#这个方法依赖ruby 2.0以上版本, 原因在 String.ord是在ruby 2.0以上才有
#在2.0以前版本可以使用 'A'.unpack('C')[0] 来获得ASCII
def compare_6(long_string, short_string)
  prime_number = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101]
  product = long_string.chars.inject(1) { |product, n| product * prime_number[ n.ord - "A".ord] }
  short_string.each_char{|c|
    return false unless product % prime_number[ c.ord - "A".ord] == 0
  }
  true
end


#这个是ruby的实现方式
def compare_ruby(long_string, short_string)
  short_string.chars - long_string.chars == []
end



if __FILE__== $0
  long = "ABCDEFGHLMNOPQRSDCGSRQPX"
  short = "DCCCCGSRQPX"
  p compare_1(long, short)
  p compare_2(long, short)
  p compare_3(long, short)
  p compare_4(long, short)
  p compare_5(long, short)
  p compare_6(long, short)
  p compare_ruby(long, short)
  puts 
  
  long = "ABCDEFGHLMNOPQRS"
  short = "DCGSRQPX"
  p compare_1(long, short)
  p compare_2(long, short)
  p compare_3(long, short)
  p compare_4(long, short)
  p compare_5(long, short)
  p compare_6(long, short)
  p compare_ruby(long, short)
end

